Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.h

```diff
@@ -0,0 +1,122 @@
+{
+    class RowFilterAdapter : public DocumentAdapter
+    {
+    public:
+        RowFilterAdapter();
+
+        void SetSourceAdapter(DocumentAdapterPtr sourceAdapter);
+
+    protected:
+        struct MatchInfoNode
+        {
+            ~MatchInfoNode()
+            {
+                for (auto child : m_childMatchState)
+                {
+                    delete child;
+                }
+                m_childMatchState.clear();
+            }
+
+            bool m_matchesSelf = false;
+            AZStd::set<MatchInfoNode*> m_matchingDescendents;
+            MatchInfoNode* m_parentNode = nullptr;
+
+            //! Deque where only row node children are populate, other children are null entries
+            AZStd::deque<MatchInfoNode*> m_childMatchState;
+
+        protected:
+            MatchInfoNode(MatchInfoNode* parentNode)
+                : m_parentNode(parentNode)
+            {
+            }
+        };
+
+        // pure virtual methods for new RowFilterAdapters
+        virtual MatchInfoNode* NewMatchInfoNode(MatchInfoNode* parentRow) const = 0;
+        virtual void CacheDomInfoForNode(const Dom::Value& domValue, MatchInfoNode* matchNode) const = 0;
+        virtual bool MatchesFilter(MatchInfoNode* matchNode) const = 0;
+
+        static bool IsRow(const Dom::Value& domValue);
+        bool IsRow(const Dom::Path& sourcePath) const;
+
+        void SetFilterActive(bool activateFilter);
+        void InvalidateFilter();
+        
+        Dom::Value GenerateContents() override;
+
+        DocumentAdapterPtr m_sourceAdapter;
+
+        void HandleReset();
+        void HandleDomChange(const Dom::Patch& patch);
+        void HandleDomMessage(const AZ::DocumentPropertyEditor::AdapterMessage& message, Dom::Value& value);
+
+        DocumentAdapter::ResetEvent::Handler m_resetHandler;
+        ChangedEvent::Handler m_changedHandler;
+        MessageEvent::Handler m_domMessageHandler;
+
+        MatchInfoNode* GetMatchNodeAtPath(const Dom::Path& sourcePath);
+
+        /*! populates the MatchInfoNode nodes for the given path and any descendent row children created.
+         *  All new nodes have their m_matchesSelf, m_matchingDescendents, and  m_matchableDomTerms set */
+        void PopulateNodesAtPath(const Dom::Path& sourcePath);
+
+        /*! updates the match states (m_matchesSelf, m_matchingDescendents) for the given row node,
+            and updates the m_matchingDescendents state for all its ancestors
+            \param rowState the row to operate on */
+        void UpdateMatchState(MatchInfoNode* rowState);
+
+        //! returns the first path in the ancestry of sourcePath that is of type Row, including self
+        Dom::Path GetRowPath(const Dom::Path& sourcePath) const;
+
+        //! indicates whether all children of a direct match are considered matching as well
+        bool m_includeAllMatchDescendents = true;
+        bool m_filterActive = false;
+
+        MatchInfoNode* m_root = nullptr;
```

Comment:
I don't see a dtor so I suspect this leaks memory (I only see a delete in the reset method). I'd suggest just making this an `AZStd::unique_ptr` (might also be worth doing with the child nodes too if we end up running into any funkiness there)

Reply:
I tried going the unique_ptr route, but since there's a deque of them to manage the children it became really unwieldy, and maybe impossible. I think the destructor covers it. Good catch!